Self type in trait implementation causes panic in #[async_trait]

Example Code:

```rust
use async_trait::async_trait;

struct Struct;

#[async_trait]
pub trait Trait {
    async fn foo();
}

#[async_trait]
impl Trait for Struct {
    async fn foo() {
        let _ = Self;
        // let _ = Struct;
    }
}
```

Error:

```
error: custom attribute panicked
  --> src/lib.rs:19:1
   |
19 | #[async_trait]
   | ^^^^^^^^^^^^^^
   |
   = help: message: called `Option::unwrap()` on a `None` value
```

I traced the error to using `Self` instead of `Struct`. non-`async` functions are not affected. Also the error is different if I change it to `Struct { x: u8 }`, but not if I cange it to `Struct(u8);`:

```
error[E0401]: can't use generic parameters from outer function
  --> src/lib.rs:22:17
   |
20 | impl Trait for Struct {
   | ---- `Self` type implicitly declared here, by this `impl`
21 |     async fn foo() {
22 |         let _ = Self { x: 0 };
   |                 ^^^^
   |                 |
   |                 use of generic parameter from outer function
   |                 use a type here instead
```

This time not a panic, but this would compile with `Struct` instead of `Self` or in a non-`async` function.

using `async-trait`0.1.19 & Rust 1.39.0

PS. thanks for this crate, it's awesome!

--- a/src/receiver.rs
+++ b/src/receiver.rs
@@ -3,8 +3,8 @@ use std::mem;
 use syn::punctuated::Punctuated;
 use syn::visit_mut::{self, VisitMut};
 use syn::{
-    Block, ExprPath, Ident, Item, Macro, Path, QSelf, Receiver, Signature, Type, TypePath,
-    WherePredicate,
+    parse_quote, Block, Error, ExprPath, Ident, Item, Macro, Path, PathArguments, QSelf, Receiver,
+    Signature, Type, TypePath, WherePredicate,
 };
 
 pub fn has_self_in_sig(sig: &mut Signature) -> bool {
@@ -84,6 +84,18 @@ impl ReplaceReceiver {
 
         let first = &path.segments[0];
         if first.ident != "Self" || !first.arguments.is_empty() {
+            return;
+        }
+
+        if path.segments.len() == 1 {
+            if let Type::Path(with) = &self.with {
+                *path = with.path.clone();
+            } else {
+                let span = first.ident.span();
+                let msg = "Self type of this impl is unsupported in expression position";
+                let error = Error::new(span, msg).to_compile_error();
+                *path = parse_quote!(::core::marker::PhantomData::<#error>);
+            }
             return;
         }
 

--- a/src/receiver.rs
+++ b/src/receiver.rs
@@ -90,6 +90,13 @@ impl ReplaceReceiver {
         if path.segments.len() == 1 {
             if let Type::Path(with) = &self.with {
                 *path = with.path.clone();
+                for segment in &mut path.segments {
+                    if let PathArguments::AngleBracketed(bracketed) = &mut segment.arguments {
+                        if bracketed.colon2_token.is_none() && !bracketed.args.is_empty() {
+                            bracketed.colon2_token = Some(Default::default());
+                        }
+                    }
+                }
             } else {
                 let span = first.ident.span();
                 let msg = "Self type of this impl is unsupported in expression position";

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -433,3 +433,22 @@ pub mod issue46 {
 
     implement_commands!(K: Send);
 }
+
+// https://github.com/dtolnay/async-trait/issues/53
+pub mod issue53 {
+    use async_trait::async_trait;
+
+    struct Struct;
+
+    #[async_trait]
+    pub trait Trait {
+        async fn foo();
+    }
+
+    #[async_trait]
+    impl Trait for Struct {
+        async fn foo() {
+            let _ = Self;
+        }
+    }
+}

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -451,4 +451,11 @@ pub mod issue53 {
             let _ = Self;
         }
     }
+
+    #[async_trait]
+    impl Trait for std::marker::PhantomData<Struct> {
+        async fn foo() {
+            let _ = Self;
+        }
+    }
 }

